File: src/bosh-monitor/lib/bosh/monitor/agent.rb
module Bosh::Monitor
  class Agent
    attr_reader   :id
    attr_reader   :discovered_at
    attr_accessor :updated_at

    ATTRIBUTES = %i[deployment job index instance_id cid job_state has_processes].freeze

    ATTRIBUTES.each do |attribute|
      attr_accessor attribute
    end

    def initialize(id, opts = {})
      raise ArgumentError, 'Agent must have an id' if id.nil?

      @id            = id
      @discovered_at = Time.now
      @updated_at    = Time.now
      @logger        = Bosh::Monitor.logger
      @intervals     = Bosh::Monitor.intervals

      @deployment = opts[:deployment]
      @job = opts[:job]
      @index = opts[:index]
      @cid = opts[:cid]
      @instance_id = opts[:instance_id]
    end

    def name
      if @deployment && @job && @instance_id
        name = "#{@deployment}: #{@job}(#{@instance_id}) [id=#{@id}, "

        name += "index=#{@index}, " if @index

        name + "cid=#{@cid}]"
      else
        state = ATTRIBUTES.each_with_object([]) do |attribute, acc|
          value = send(attribute)
          acc << "#{attribute}=#{value}" if value
        end

        "agent #{@id} [#{state.join(', ')}]"
      end
    end

    def timed_out?
      (Time.now - @updated_at) > @intervals.agent_timeout
    end

    def rogue?
      (Time.now - @discovered_at) > @intervals.rogue_agent_alert && @deployment.nil?
    end

    def is_not_running?
      @job_state.to_s != 'running' || @has_processes == false
    end

    def update_instance(instance)
      @job = instance.job
      @index = instance.index
      @cid = instance.cid
      @instance_id = instance.id
      @job_state = instance.job_state
      @has_processes = instance.has_processes
    end
  end
end

File: src/bosh-monitor/pkg/instance/agent.go
package instance

import (
	"fmt"
	"time"
)

type Agent struct {
	AgentID           string
	Deployment        string
	Job               string
	Index             string
	InstanceID        string
	CID               string
	DiscoveredAt      time.Time
	UpdatedAt         time.Time
	JobState          string
	NumberOfProcesses int

	agentTimeout    time.Duration
	rogueAgentAlert time.Duration
}

func NewAgent(id string, opts ...AgentOption) *Agent {
	a := &Agent{
		AgentID:         id,
		DiscoveredAt:    time.Now(),
		UpdatedAt:       time.Now(),
		agentTimeout:    60 * time.Second,
		rogueAgentAlert: 120 * time.Second,
	}
	for _, opt := range opts {
		opt(a)
	}
	return a
}

type AgentOption func(*Agent)

func WithDeployment(d string) AgentOption {
	return func(a *Agent) { a.Deployment = d }
}

func WithAgentTimeout(d time.Duration) AgentOption {
	return func(a *Agent) { a.agentTimeout = d }
}

func WithRogueAgentAlert(d time.Duration) AgentOption {
	return func(a *Agent) { a.rogueAgentAlert = d }
}

func (a *Agent) Name() string {
	if a.Deployment != "" && a.Job != "" && a.InstanceID != "" {
		name := fmt.Sprintf("%s: %s(%s) [id=%s, ", a.Deployment, a.Job, a.InstanceID, a.AgentID)
		if a.Index != "" {
			name += fmt.Sprintf("index=%s, ", a.Index)
		}
		name += fmt.Sprintf("cid=%s]", a.CID)
		return name
	}

	var parts []string
	if a.Deployment != "" {
		parts = append(parts, fmt.Sprintf("deployment=%s", a.Deployment))
	}
	if a.Job != "" {
		parts = append(parts, fmt.Sprintf("job=%s", a.Job))
	}
	if a.Index != "" {
		parts = append(parts, fmt.Sprintf("index=%s", a.Index))
	}
	if a.CID != "" {
		parts = append(parts, fmt.Sprintf("cid=%s", a.CID))
	}
	if a.InstanceID != "" {
		parts = append(parts, fmt.Sprintf("instance_id=%s", a.InstanceID))
	}

	state := ""
	for i, p := range parts {
		if i > 0 {
			state += ", "
		}
		state += p
	}
	return fmt.Sprintf("agent %s [%s]", a.AgentID, state)
}

func (a *Agent) TimedOut() bool {
	return time.Since(a.UpdatedAt) > a.agentTimeout
}

func (a *Agent) Rogue() bool {
	return time.Since(a.DiscoveredAt) > a.rogueAgentAlert && a.Deployment == ""
}

func (a *Agent) UpdateInstance(inst *Instance) {
	a.Job = inst.Job
	a.Index = inst.Index
	a.CID = inst.CID
	a.InstanceID = inst.InstanceID
}
